src/mcp/shared/proxy.py

```diff
@@ -0,0 +1,144 @@
+from contextlib import asynccontextmanager
+
+import anyio
+from anyio.streams.memory import MemoryObjectReceiveStream, MemoryObjectSendStream
+
+from mcp.shared.message import SessionMessage
+
+logger = logging.getLogger(__name__)
+
+MessageStream = tuple[
+    MemoryObjectReceiveStream[SessionMessage | Exception],
+    MemoryObjectSendStream[SessionMessage],
+]
+
+
+async def _handle_error(
+    error: Exception,
+    onerror: Callable[[Exception], None | Awaitable[None]] | None,
+) -> None:
+    """Handle an error by calling the error callback if provided."""
+    if onerror:
+        try:
+            result = onerror(error)
+            if isinstance(result, Awaitable):
+                await result
+        except Exception as callback_error:  # pragma: no cover
+            logger.exception("Error in onerror callback", exc_info=callback_error)
+
+
+async def _forward_message(
+    message: SessionMessage | Exception,
+    write_stream: MemoryObjectSendStream[SessionMessage],
+    onerror: Callable[[Exception], None | Awaitable[None]] | None,
+    source: str,
+) -> None:
+    """Forward a single message, handling exceptions appropriately."""
+    if isinstance(message, SessionMessage):
+        await write_stream.send(message)
+    else:
+        # message is Exception (type narrowing)
+        logger.debug(f"Exception received from {source}: {message}")
+        await _handle_error(message, onerror)
+        # Exceptions are not forwarded as messages (write streams only accept SessionMessage)
+
+
+async def _forward_loop(
+    read_stream: MemoryObjectReceiveStream[SessionMessage | Exception],
+    write_stream: MemoryObjectSendStream[SessionMessage],
+    onerror: Callable[[Exception], None | Awaitable[None]] | None,
+    source: str,
+) -> None:
+    """Forward messages from read_stream to write_stream."""
+    try:
+        async with read_stream:
+            async for message in read_stream:
+                try:
+                    await _forward_message(message, write_stream, onerror, source)
+                except anyio.ClosedResourceError:
+                    logger.debug(f"{source} write stream closed")
+                    break
+                except Exception as exc:  # pragma: no cover
+                    # This covers non-ClosedResourceError exceptions during message forwarding
+                    # (e.g., from custom stream implementations)
+                    logger.exception(f"Error forwarding message from {source}", exc_info=exc)
+                    await _handle_error(exc, onerror)
+    except anyio.ClosedResourceError:
+        logger.debug(f"{source} read stream closed")
+    except Exception as exc:  # pragma: no cover
+        # This covers exceptions during stream iteration setup
+        # (e.g., from custom stream implementations)
+        logger.exception(f"Error in forward loop from {source}", exc_info=exc)
+        await _handle_error(exc, onerror)
```

Comment:
I think it would be a bit cleaner if this `_handle_error` runs on an `__exit__` from a context manager that we can create to run in this function?

Also, can we drop the `logger.exception` and `logger.debug` from everywhere?